inst/include/distributions/functors/lognormal_lpdf.hpp

```diff
@@ -1,74 +1,94 @@
+        virtual const Type evaluate(const bool &do_log)
+        {
+            this->mu.resize(this->observed_values.size());
+            this->sd.resize(this->observed_values.size());
+            is_na.resize(this->observed_values.size());
+            for (size_t i = 0; i < this->expected_values.size(); i++)
+            {
+                if (this->expected_values.size() == 1)
+                {
+                    this->mu[i] = this->expected_values[0];
+                }
+                else
+                {
+                    this->mu[i] = this->expected_values[i];
+                }
 
-    virtual const Type evaluate(){
-        this->mu.resize(this->observed_value.size());
-        this->sd.resize(this->observed_value.size());
-        for(int i=0; i<this->expected_value.size(); i++){
-            if(this->expected_value.size() == 1){
-                this->mu[i] = this->expected_value[0];
-            } else {
-                this->mu[i] = this->expected_value[i];
-            }
-       
-            if(log_sd.size() == 1){
-                sd[i] = exp(log_sd[0]);
-            } else {
-                sd[i] = exp(log_sd[i]);
-            }
-        }
-        this->nll_vec.resize(this->observed_value.size());
-        for(int i=0; i<this->observed_value.size(); i++){
-           // this->nll_vec[i] = this->keep[i] * -dnorm(this->observed_value[i], mu[i], sd[i], true);
-            this->nll_vec[i] = -dnorm(log(this->observed_value[i]), mu[i], sd[i], true) - log(this->observed_value[i]);
-            nll += this->nll_vec[i];
-            if(this->simulate_flag){
-                FIMS_SIMULATE_F(this->of){ //preprocessor definition in interface.hpp
-                //this simulates data that is mean biased
-                    this->observed_value[i] = exp(rnorm(mu[i], sd[i]));
+                if (log_sd.size() == 1)
+                {
+                    sd[i] = fims_math::exp(log_sd[0]);
+                }
+                else
+                {
+                    sd[i] = fims_math::exp(log_sd[i]);
                 }
-                
             }
-            
-          /* osa not working yet
-            if(osa_flag){//data observation type implements osa residuals
-                //code for osa cdf method
-                this->nll_vec[i] = this->keep.cdf_lower[i] * -log( pnorm(this->observed_value[i], mu[i], sd[i]) );
-                this->nll_vec[i] = this->keep.cdf_upper[i] * -log( 1.0 - pnorm(this->observed_value[i], mu[i], sd[i]) );
-            } */
+            this->nll_vec.resize(this->observed_values.size());
+            for (size_t i = 0; i < this->observed_values.size(); i++)
+            {
+              if(!is_na[i]){
+                #ifdef TMB_MODEL
+                // this->nll_vec[i] = this->keep[i] * -dnorm(this->observed_values[i], mu[i], sd[i], do_log);
+                this->nll_vec[i] = -dnorm(log(this->observed_values[i]), mu[i], sd[i], true) + log(this->observed_values[i]);
```

Comment:
Adding the log of observed values to the nll makes complete sense when the observed are fixed data. Where this can run into problems is when the "observed" are actually derived quantities/estimates, such as when stock-recruitment residuals are assumed to follow a lognormal distribution around the stock-recruitment curve. In the latter case, adding the log of the "observed" values gives the minimizer an incentive to estimate lower "observed" values. Jon Deroba and Tim Miller wrote a paper about this: https://doi.org/10.1016/j.fishres.2015.12.002. No need to change anything here, just a reminder to be careful about how the lognormal distribution is used in FIMS.

Reply:
Thanks for noting this. We will be tracking in the model whether or not input to the distributions are data. We could add a conditional that only includes the constant term when its a data nll.